File: java/patterns/src/main/java/org/prototypepatterns/behavioral/command/CommandPattern.java
package org.prototypepatterns.behavioral.command;

public final class CommandPattern {
    private CommandPattern() {
    }

    interface Command {
        String execute();
    }

    static final class Light {
        String on() {
            return "light:on";
        }
    }

    static final class LightOnCommand implements Command {
        private final Light light;

        LightOnCommand(Light light) {
            this.light = light;
        }

        @Override
        public String execute() {
            return light.on();
        }
    }

    static final class RemoteButton {
        private final Command command;

        RemoteButton(Command command) {
            this.command = command;
        }

        String press() {
            return command.execute();
        }
    }

    public static String demo() {
        return new RemoteButton(new LightOnCommand(new Light())).press();
    }
}

File: csharp/patterns/src/PrototypePatterns/Behavioral/Command/CommandPattern.cs
namespace PrototypePatterns.Behavioral.Command;

public interface ICommand
{
    string Execute();
}

public sealed class Light
{
    public string On() => "light:on";
}

public sealed class LightOnCommand(Light light) : ICommand
{
    public string Execute() => light.On();
}

public sealed class RemoteButton(ICommand command)
{
    public string Press() => command.Execute();
}

public static class CommandPattern
{
    public static string Demo() => new RemoteButton(new LightOnCommand(new Light())).Press();
}
